internal/core/metadata/application/deviceprofile.go

```diff
@@ -206,6 +207,126 @@ func PatchDeviceProfileBasicInfo(ctx context.Context, dto dtos.UpdateDeviceProfi
 	return nil
 }
 
+func DeleteDeviceResourceByName(profileName string, resourceName string, dic *di.Container) errors.EdgeX {
+	if profileName == "" {
+		return errors.NewCommonEdgeX(errors.KindContractInvalid, "profile name is empty", nil)
+	}
+	if resourceName == "" {
+		return errors.NewCommonEdgeX(errors.KindContractInvalid, "resource name is empty", nil)
+	}
+
+	strictProfileChanges := container.ConfigurationFrom(dic.Get).Writable.ProfileChange.StrictDeviceProfileChanges
+	if strictProfileChanges {
+		return errors.NewCommonEdgeX(errors.KindServiceLocked, "profile change is not allowed", nil)
+	}
+
+	dbClient := container.DBClientFrom(dic.Get)
+	// Check the associated Device and ProvisionWatcher existence
+	devices, err := dbClient.DevicesByProfileName(0, 1, profileName)
+	if err != nil {
+		return errors.NewCommonEdgeXWrapper(err)
+	} else if len(devices) > 0 {
+		return errors.NewCommonEdgeX(errors.KindStatusConflict, "fail to delete the device profile when associated device exists", nil)
+	}
+
+	provisionWatchers, err := dbClient.ProvisionWatchersByProfileName(0, 1, profileName)
+	if err != nil {
+		return errors.NewCommonEdgeXWrapper(err)
+	} else if len(provisionWatchers) > 0 {
+		return errors.NewCommonEdgeX(errors.KindStatusConflict, "fail to delete the device profile when associated provisionWatcher exists", nil)
+	}
+
+	profile, err := dbClient.DeviceProfileByName(profileName)
+	if err != nil {
+		return errors.NewCommonEdgeXWrapper(err)
+	}
+
+	index := -1
+	for i := range profile.DeviceResources {
+		if profile.DeviceResources[i].Name == resourceName {
+			index = i
+			break
+		}
+	}
+	if index == -1 {
+		return errors.NewCommonEdgeX(errors.KindEntityDoesNotExist, "device resource not found", nil)
+	}
+
+	profile.DeviceResources = append(profile.DeviceResources[:index], profile.DeviceResources[index+1:]...)
+	profileDTO := dtos.FromDeviceProfileModelToDTO(profile)
+	e := (&profileDTO).Validate()
+	if e != nil {
+		return errors.NewCommonEdgeXWrapper(e)
+	}
+
+	err = dbClient.UpdateDeviceProfile(profile)
+	if err != nil {
+		return errors.NewCommonEdgeXWrapper(err)
+	}
+
+	return nil
+}
+
+func DeleteDeviceCommandByName(profileName string, commandName string, dic *di.Container) errors.EdgeX {
+	if profileName == "" {
+		return errors.NewCommonEdgeX(errors.KindContractInvalid, "profile name is empty", nil)
+	}
+	if commandName == "" {
+		return errors.NewCommonEdgeX(errors.KindContractInvalid, "command name is empty", nil)
+	}
+
+	strictProfileChanges := container.ConfigurationFrom(dic.Get).Writable.ProfileChange.StrictDeviceProfileChanges
+	if strictProfileChanges {
+		return errors.NewCommonEdgeX(errors.KindServiceLocked, "profile change is not allowed", nil)
```

Comment:
`profile change is not allowed when StrictDeviceProfileChanges config is enabled`